src/Task/TaskScheduler.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Phpcq\Task;
+
+use Generator;
+use Phpcq\Exception\Exception;
+use Phpcq\PluginApi\Version10\OutputInterface;
+use Phpcq\PluginApi\Version10\RuntimeException as PluginApiRuntimeException;
+use Phpcq\PluginApi\Version10\Task\ParallelTaskInterface;
+use Phpcq\PluginApi\Version10\Task\ReportWritingTaskInterface;
+use Phpcq\PluginApi\Version10\ToolReportInterface;
+use Phpcq\Report\Report;
+use SplObjectStorage;
+
+class TaskScheduler
+{
+    public const LOG_START  = '%1$s starting';
+    public const LOG_FAILED = '%1$s failed';
+    public const LOG_END    = '%1$s finished';
+
+    /**
+     * @var ReportWritingTaskInterface[]|Generator
+     * @psalm-var Generator<array-key, ReportWritingTaskInterface>
+     */
+    private $tasks;
+
+    /** @var int */
+    private $parallelThreads;
+
+    /** @var OutputInterface */
+    private $output;
+
+    /** @var Report */
+    private $report;
+
+    /** @var bool */
+    private $fastFinish;
+
+    /** @var bool */
+    private $stop;
+
+    /** @var bool */
+    private $success;
+
+    /** @var int */
+    private $runningThreads;
+
+    /** @var SplObjectStorage<ParallelTaskInterface, ToolReportInterface> */
+    private $threads;
+
+    public function __construct(
+        TasklistInterface $tasks,
+        int $parallelThreads,
+        Report $report,
+        OutputInterface $output,
+        bool $fastFinish
+    ) {
+        $this->tasks           = $this->nextTaskGenerator($tasks);
+        $this->parallelThreads = $parallelThreads;
+        $this->report          = $report;
+        $this->output          = $output;
+        $this->fastFinish      = $fastFinish;
+        $this->stop            = false;
+        $this->success         = true;
+        $this->runningThreads  = 0;
+        /** @psalm-suppress PropertyTypeCoercion - the empty object storage IS the correct type. */
+        $this->threads = new SplObjectStorage();
+    }
+
+    public function run(): bool
+    {
+        $this->fillUp();
```

Comment:
We should reset state here or disallow multiple calls to avoid unexpectable behaviour.

Reply:
Addressed in f8a95fe & 317ef23